astropy/nddata/mixins/ndarithmetic.py

```diff
@@ -24,211 +30,588 @@ class NDArithmeticMixin(object):
-            if operand.uncertainty:
-                operand_uncert_value = operand.unit.to(self.unit,
-                                                       operand_uncert_value)
+                kwargs['meta'] = self.meta
         else:
-            operand_data = operand.data
+            kwargs['meta'] = self._arithmetic_meta(operation, operand,
+                                        meta_kwds_operate, meta_kwds_set)
+
+        # Wrap the individual results into a new instance of the same class.
+        return result, kwargs
+
+    def _arithmetic_data(self, operation, operand):
+        """
+        Calculate the resulting data
 
-        if operand.uncertainty:
-            # Create a copy here in case this is returned as the uncertainty
-            # of the result.
-            operand_uncertainty = \
-                operand.uncertainty.__class__(operand_uncert_value, copy=True)
+        Parameters
+        ----------
+        operation: `str`
+            see `NDArithmeticMixin.arithmetic` parameter description.
+
+        operand: `NDData`-like instance
+            The second operand wrapped in an instance of the same class as
+            self.
+
+        decompose_unit: `bool`
+            see :meth:`NDArithmeticMixin.arithmetic` parameter description.
+
+        Returns
+        -------
+        result_data: `~numpy.ndarray` or `~astropy.units.Quantity`
+            if both operands had no unit the resulting data is a simple numpy
+            array, but if any of the operands had a unit the return is a
+            Quantity.
+        """
+
+        # Find the right function for that operation... (numpy ufuncs)
+        if operation == 'addition':
+            operator = np.add
+        elif operation == 'subtraction':
+            operator = np.subtract
+        elif operation == 'multiplication':
+            operator = np.multiply
+        elif operation == 'division':
+            # Unfortunatly the future import does not implicitly call the
+            # true division of numpy so explicitly call np.true_divide instead
+            # of np.divide (until python2 support is dropped ... just joking)
+            operator = np.true_divide
+        else:
+            raise ValueError('Unsupported operation')
+
+        # Do the calculation with or without units
+        if self.unit is None and operand.unit is None:
+            result = operator(self.data, operand.data)
+        elif self.unit is None:
+            result = operator(self.data * dimensionless_unscaled,
+                              operand.data * operand.unit)
+        elif operand.unit is None:
+            result = operator(self.data * self.unit,
+                              operand.data * dimensionless_unscaled)
         else:
-            operand_uncertainty = None
+            result = operator(self.data * self.unit,
+                              operand.data * operand.unit)
+
+        return result
```

Comment:
This might all benefit by some kind of operator class like I describe in #3861.  That was specifically in the context of compound models.  But maybe it could have a broader use than that, such as handling on a per-operator basis how to handle uncertainties for that operator, etc.  Not 100% clear if it would help here since I haven't done a deep dive on this yet but it's just a thought.